ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,530 @@
+{
+    if (_ebpf_netevent_event_hook_provider_context) {
+        ebpf_extension_hook_provider_unregister(_ebpf_netevent_event_hook_provider_context);
+        _ebpf_netevent_event_hook_provider_context = NULL;
+    }
+    if (_ebpf_netevent_event_program_info_provider_context) {
+        ebpf_extension_program_info_provider_unregister(_ebpf_netevent_event_program_info_provider_context);
+        _ebpf_netevent_event_program_info_provider_context = NULL;
+    }
+    if (_event_buffer) {
+        ExFreePool(_event_buffer);
+        _event_buffer = NULL;
+    }
+}
+
+//
+// eBPF NetEvent Program Information NPI helper routines.
+//
+static ebpf_result_t
+_ebpf_netevent_program_context_create(
+    _In_reads_bytes_opt_(data_size_in) const uint8_t* data_in,
+    size_t data_size_in,
+    _In_reads_bytes_opt_(context_size_in) const uint8_t* context_in,
+    size_t context_size_in,
+    _Outptr_ void** context)
+{
+    EBPF_EXT_LOG_ENTRY();
+    ebpf_result_t result;
+    netevent_event_md_t* netevent_event_context = NULL;
+
+    *context = NULL;
+
+    if (context_in == NULL || context_size_in < sizeof(netevent_event_md_t)) {
+        EBPF_EXT_LOG_MESSAGE(EBPF_EXT_TRACELOG_LEVEL_ERROR, EBPF_EXT_TRACELOG_KEYWORD_NETEVENT, "Context is required");
+        result = EBPF_INVALID_ARGUMENT;
+        goto Exit;
+    }
+
+    // Allocate memory for the context.
+    netevent_event_context = (netevent_event_md_t*)ExAllocatePoolUninitialized(
+        NonPagedPoolNx, sizeof(netevent_event_md_t), EBPF_NETEVENT_EXTENSION_POOL_TAG);
+    EBPF_EXT_BAIL_ON_ALLOC_FAILURE_RESULT(
+        EBPF_EXT_TRACELOG_KEYWORD_NETEVENT, netevent_event_context, "netevent_event_context", result);
+
+    // Copy the context from the caller.
+    memcpy(netevent_event_context, context_in, sizeof(netevent_event_md_t));
+
+    // Copy the event's pointer & size from the caller, to the out context.
+    netevent_event_context->event_data_start = (uint8_t*)data_in;
+    netevent_event_context->event_data_end = (uint8_t*)data_in + data_size_in;
+    *context = netevent_event_context;
+    netevent_event_context = NULL;
+    result = EBPF_SUCCESS;
+
+Exit:
+    if (netevent_event_context) {
+        ExFreePool(netevent_event_context);
+        netevent_event_context = NULL;
+    }
+    EBPF_EXT_RETURN_RESULT(result);
+}
+
+static void
+_ebpf_netevent_program_context_destroy(
+    _In_opt_ void* context,
+    _Out_writes_bytes_to_(*data_size_out, *data_size_out) uint8_t* data_out,
+    _Inout_ size_t* data_size_out,
+    _Out_writes_bytes_to_(*context_size_out, *context_size_out) uint8_t* context_out,
+    _Inout_ size_t* context_size_out)
+{
+    EBPF_EXT_LOG_ENTRY();
+
+    netevent_event_md_t* netevent_event_context = (netevent_event_md_t*)context;
+    netevent_event_md_t* netevent_event_context_out = (netevent_event_md_t*)context_out;
+
+    if (!netevent_event_context) {
+        goto Exit;
+    }
+
+    if (context_out != NULL && *context_size_out >= sizeof(netevent_event_md_t)) {
```

Comment:
Do we need to check that context_size_out is not NULL?

Reply:
Similarly, do we need to check data_size_out for non-null as well?

Reply:
Why are you asking this?

Reply:
On my read, we would crash if context_size_out of data_size_out is NULL. We should probably add a NULL check to ensure that we don't.

Reply:
The [`_Out_writes_bytes_to_`](https://learn.microsoft.com/en-us/cpp/code-quality/annotating-function-parameters-and-return-values?view=msvc-170) annotation specifies that `data_out` and `context_out` will be written to with a number of bytes equal to the value pointed to by `data_size_out` and `context_size_out`, respectively. This implies that the pointers `data_size_out` and `context_size_out` must be valid (i.e., not null) because their values are used to determine how much data is written to `data_out` and `context_out`. If not, the static analysis should trigger a warning, which we treat as error by policy.

Nothing prevents to adding further checks (i.e. including in the rest of the codebase), but we use SAL to simplify the code.